iwyu.cc

```diff
@@ -1631,7 +1625,8 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // The comment, if not nullptr, is extra text that is included along
   // with the warning message that iwyu emits.
   virtual void ReportTypeUse(SourceLocation used_loc, const Type* type,
-                             const char* comment = nullptr) {
+                             const char* comment = nullptr,
+                             const set<const Type*>& blocked_types = {}) {
```

Comment:
I'm not a huge fan of the `blocked_types` argument, because it smells of a control flag -- something the caller passes in only to modify the control flow of the function. My instinct is to either:

* move the checking of `type` vs. `blocked_types` out if this function and into the few callers where it matters, or
* resolve `blocked_types` (or "provided types") inside `ReportTypeUse`

But I realize that neither is probably very straightforward... Let me see if I can get my head around the interdependencies and figure something out.

Reply:
On that note, it's also a little suspicious that `ReportTypesUse` does not do anything with these provided types. I'm guessing it's because the filtering happens before types end up in the cache whose data is passed to `ReportTypesUse`. But that creates a strange implicit dependency between the cache and the type reporting. E.g. what if I wanted to use `ReportTypesUse` to report a set of types that _do_ need provision filtering? The API implies that `ReportTypesUse` is just a batch version of `ReportTypeUse`.

Reply:
As an alternative, I can suggest moving `blocked_types_` field from `InstantiatedTemplateVisitor` into `IwyuBaseAstVisitor` as a protected field. `IwyuBaseAstVisitor::ReportTypeUse` function would then fill in the set when it finds a typedef and then restore its state.

Reply:
That also feels a little funny. Let's think on this for a wile and see if a third alternative comes up.

Reply:
> move the checking of `type` vs. `blocked_types` out if this function and into the few callers where it matters

The problem here is that `ReportTplSpecComponentTypes` analysis should be launched even when template specialization type is "blocked" (see [the comment](https://github.com/include-what-you-use/include-what-you-use/pull/1248/commits/7ac841682320b15b313edd65bca71ad655823f1c#diff-7ca107c8b836c1765e8b735b74f06833bec48f308a76017886cee1d5736002d9R1687)). Hence, `ReportTypeUse` should be called even when the type is blocked.

A virtual (or non-virtual using CRTP) function `GetBlockedTypes()` may be introduced which would return `blocked_types_` for `InstantiatedTemplateVisitor` and an empty set for `IwyuAstConsumer`. But how `IwyuBaseAstVisitor::ReportTypeUse` could transfer `provided_types` of a typedef between its recursive calls if not with an argument or class state? I can suggest moving its body into private function `IwyuBaseAstVisitor::ReportTypeUseInternal` which would take `blocked_type` parameter and its public (for subclasses) interface `ReportTypeUse` would not.

Reply:
> I can suggest moving its body into private function IwyuBaseAstVisitor::ReportTypeUseInternal which
> would take blocked_type parameter and its public (for subclasses) interface ReportTypeUse would not.

That actually sounds nicer to me.

`ReportTypeUse` would then get the type components and pass them as "blocked" as necessary? That would also nicely solve the `ReportTypesUse` asymmetry.

Reply:
(please add it first as a separate commit on top, so we can change our minds if it somehow turns out worse)